spark/src/main/java/org/apache/iceberg/spark/actions/RepairManifestHelper.java

```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  /**
+   * Diffs two DataFile for potential for repair
+   * @return a set of fields in human-readable format that differ between these DataFiles
+   */
+  static Set<String> diff(DataFile first, DataFile second) {
+    Set<String> result = new HashSet<>();
+    if (first.fileSizeInBytes() != second.fileSizeInBytes()) {
+      result.add("file_size_in_bytes");
+    }
+    if (first.recordCount() != second.recordCount()) {
+      result.add("record_count");
+    }
+    if (!Objects.equals(first.columnSizes(), second.columnSizes())) {
+      result.add("column_sizes");
+    }
+    if (!Objects.equals(first.valueCounts(), second.valueCounts())) {
+      result.add("value_counts");
+    }
+    if (!Objects.equals(first.nullValueCounts(), second.nullValueCounts())) {
+      result.add("null_value_counts");
+    }
+    if (!Objects.equals(first.nanValueCounts(), second.nanValueCounts())) {
+      result.add("nan_value_counts");
+    }
+    if (!Objects.equals(first.lowerBounds(), second.lowerBounds())) {
+      result.add("lower_bounds");
+    }
+    if (!Objects.equals(first.upperBounds(), second.upperBounds())) {
+      result.add("upper_bounds");
+    }
+    return result;
+  }
+
+  /**
+   * Given a data file pointer, return a repaired version if actual file information does not match.
+   * @param file spark data file
+   * @param spec user-specified spec
+   * @param table table information
+   * @param conf Hadoop configuration
+   * @return A repaired DataFile if repair was done (file information did not match), or None if not
+   */
+  static Optional<RepairedDataFile> repairDataFile(SparkDataFile file,
+                                                   Table table,
+                                                   PartitionSpec spec,
+                                                   Configuration conf) {
+    DataFiles.Builder newDfBuilder = DataFiles.builder(spec).copy(file);
+    Path path = new Path(file.path().toString());
+    try {
+      FileSystem fs = path.getFileSystem(conf);
+      FileStatus status = fs.getFileStatus(path);
+      newDfBuilder.withStatus(status);
+      String nameMappingString = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+      NameMapping nameMapping = nameMappingString != null ? NameMappingParser.fromJson(nameMappingString) : null;
+      newDfBuilder.withMetrics(getMetrics(file.format(), status, conf,
+          MetricsConfig.fromProperties(table.properties()), nameMapping));
+
+      DataFile newFile = newDfBuilder.build();
+      Set<String> diff = RepairManifestHelper.diff(file, newFile);
+      if (diff.isEmpty()) {
+        return Optional.empty();
+      } else {
+        return Optional.of(new RepairedDataFile(newFile, diff));
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  /**
+   * Represents a repaired DataFile
+   */
+  public static class RepairedDataFile {
```

Comment:
Needs to be public for Spark serialization

Reply:
Maybe put this comment into class's comment as well, "Needs to be public for Spark serialization"?